### DISCUSSION TITLE
[Proof-of-Concept] Pydantic before runtime: static conversion to/fro SQLalchemy, argparse, JSON schema

### DISCUSSION THREAD

**Discussion by SamuelMarks** [pydantic/pydantic#4816] [Projects using Pydantic]
### Initial Checks

- [X] I have searched Google & GitHub for similar requests and couldn't find anything
- [X] I have read and followed [the docs](https://pydantic-docs.helpmanual.io) and still think this feature is missing

### Description

It would be great if you could use Pydantic at a purely static code analysis level (i.e., before compile time / runtime).

To take the FastAPI use-case, you should be able to produce OpenAPI config without actually running the FastAPI server; i.e., just by statically analysis the codebase.

…so I just jerryrigged my project to support Pydantic (FastAPI next!):
**https://github.com/offscale/cdd-python**
[100% test and doc coverage, support for last 6 versions of Python]

Future work:
  - Add missing Pydantic features that aren't supported; it currently just treats Pydantic `class`es as normal Python 
  - `class`es… but it should be straightforward to extend with greater support.
  - SQLalchemy 2.0 support
  - FastAPI support
  - …other framework support (e.g., other ORMs, other route handlers like Flask)
  - Finish CST implementation and improve type inferrer so this can be used on major codebases to insert type annotations throughout

Feedback is most welcome

### Affected Components

- [ ] [Compatibility between releases](https://pydantic-docs.helpmanual.io/changelog/)
- [X] [Data validation/parsing](https://pydantic-docs.helpmanual.io/usage/models/#basic-model-usage)
- [X] [Data serialization](https://pydantic-docs.helpmanual.io/usage/exporting_models/) - `.dict()` and `.json()`
- [X] [JSON Schema](https://pydantic-docs.helpmanual.io/usage/schema/)
- [X] [Dataclasses](https://pydantic-docs.helpmanual.io/usage/dataclasses/)
- [ ] [Model Config](https://pydantic-docs.helpmanual.io/usage/model_config/)
- [X] [Field Types](https://pydantic-docs.helpmanual.io/usage/types/) - adding or changing a particular data type
- [ ] [Function validation decorator](https://pydantic-docs.helpmanual.io/usage/validation_decorator/)
- [X] [Generic Models](https://pydantic-docs.helpmanual.io/usage/models/#generic-models)
- [ ] [Other Model behaviour](https://pydantic-docs.helpmanual.io/usage/models/) - `construct()`, pickling, private attributes, ORM mode
- [X] [Plugins](https://pydantic-docs.helpmanual.io/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

**Comment by samuelcolvin** [comment]
Looks really interesting, thanks for posting.

It reminds me somewhat of a feature of plotly I saw demoed a few years ago (no idea if it's still available) where you could take code to create a given plot and automatically change the language it's "written" in.

One piece of feedback, it might help to provide some example use cases to better demonstrate the uses of your tool.

Also worth noting that static analysis like this can't exactly duplicate the schema or behaviour your would get from a pydantic model due to validators adding custom logic to validation and `__modify_schema__` potentially customising schema generation.

**Reply by SamuelMarks** [reply]
I hear you. I've been working on a suite of compilers to automatically translate—within and—between:

| Language             | Compiler                                               |
| -------------------- | ------------------------------------------------------ |
| Python               | [cdd-python](https://github.com/offscale/cdd-python)   |
| C                    | [cdd-c](https://github.com/SamuelMarks/cdd-c)          |
| Java (Android)       | [cdd-java](https://github.com/offscale/cdd-java)       |
| Kotlin (Android)     | [cdd-kotlin](https://github.com/offscale/cdd-kotlin)   |
| Swift (iOS)          | [cdd-swift](https://github.com/offscale/cdd-swift-ios) |
| TypeScript (Angular) | [cdd-ts-ng](https://github.com/offscale/cdd-ts-ng)     |
| Rust                 | [cdd-rust](https://github.com/offscale/cdd-rust)       |

…but you're also right, it doesn't handle the fancy features of the libraries it "supports"; e.g., [`__modify_schema__`](https://docs.pydantic.dev/usage/schema/#modifying-schema-in-custom-fields)

Looking through the FastAPI docs they have some pretty awful duplication recommended for SQLalchemy and pydantic in the one codebase: https://fastapi.tiangolo.com/tutorial/sql-databases/

…now, I'm not interested in writing my own "better" framework. Because chances are, no one—'cept me!—will use it. But I can automate that duplication with my [cdd-python](https://pypi.org/project/python-cdd) project. Cum SQLalchemy 2 and I wouldn't be surprised if FastAPI has examples sans-Pydantic; as both have PEP484 type annotations. Or I guess if I really get into FastAPI then I can replace their OpenAPI generating logic with [cdd-python](https://pypi.org/project/python-cdd)'s…

PS: My project is probably too small-time to be referenced https://docs.pydantic.dev/#using-pydantic but it's your call.
